Merge dash renderer with DashPy

HammadTheOne opened the pull request:
This PR addresses plotly/dash-core#225 as part of the Dash monorepo project, and merges the `dash-renderer` artifacts within Dash. Dependencies and resources are loaded from Dash itself, which will make the release and  development process a bit easier. This PR also introduces `gulpfile` to the Dash repo as a tool for moving and modifying files/directories.

## Contributor Checklist

- [x] I have broken down my PR scope into the following TODO tasks
   -  [x] Merge dash-renderer artifacts
   -  [x] DashPy __init__.py will need to be aware of the dynamically defined resources
   -  [x] DashPy needs to get renderer resources from itself instead of dash_renderer 
   -  [x] ~~Set up gulpfile to automate renderer updates merge to dash~~
   -  [x] Fix following glob sequence in `MANIFEST`: `include dash/dash-renderer/**/**/**/**/**` (Fixed in 9900303)
   -  [x] Add cleanup job for `node_modules` (Fixed in 317d7e8d)
   -  [x] Change build folder to `current_directory` .

- [x] I have run the tests locally and they passed. (refer to testing section in [contributing](https://github.com/plotly/dash/blob/master/CONTRIBUTING.md))
- [ ] I have added tests, or extended existing tests, to cover any new features or bugs fixed in this PR

### optionals

- [x] I have added entry in the `CHANGELOG.md`
- [ ] If this PR needs a follow-up in **dash docs**, **community thread**, I have mentioned the relevant URLS as follows
    -  [ ] this GitHub [#PR number]() updates the dash docs
    -  [ ] here is the show and tell thread in Plotly Dash community
    
Closes plotly/dash-core#225

alexcjohnson commented:
@HammadTheOne the final renderer build in `dash/deps` looks great, but I wonder about the process - can't we just use the same process we already had - [`renderer build`](https://github.com/plotly/dash/blob/edce37426470b900a0081969af30364e43c9a888/dash-renderer/package.json#L17) that [calls](https://github.com/plotly/dash/blob/edce37426470b900a0081969af30364e43c9a888/setup.py#L38) out to [`build_process`](https://github.com/plotly/dash/blob/dev/dash/development/build_process.py), as well as the [webpack output](https://github.com/plotly/dash/blob/edce37426470b900a0081969af30364e43c9a888/dash-renderer/webpack.config.js#L44) - and point it at `dash/deps` instead of `dash-renderer/dash_renderer`?

My rationale here is that as you have it here we're duplicating files: the ones actually getting used by dash are going to be harder to keep updated during renderer development because they need to go through an extra step, and the ones being created inside the renderer directory never get directly used at all. Also `dash_renderer/__init__.py` (based on templating from [`init.template`](https://github.com/plotly/dash/blob/dev/dash-renderer/init.template)) is a little cleaner than the new `_dash_renderer.py`, as it has all the versions inlined.

Finally (and again this is thinking about development in the renderer itself) since the entire `dash/deps` directory is build artifacts based on source files elsewhere in the repo, it should be gitignored except in the released (`master`) branch.

HammadTheOne commented:
> @HammadTheOne the final renderer build in dash/deps looks great, but I wonder about the process - can't we just use the same process we already had - renderer build that calls out to build_process, as well as the webpack output - and point it at dash/deps instead of dash-renderer/dash_renderer?

I think that makes a lot of sense. I've removed the `gulpfile` method entirely, and modified the existing build process to make it a little more flexible and to update and source dependencies directly to and from the `deps` directory. A couple of things to note - in the existing [ `build_process` call ](https://github.com/plotly/dash/blob/20b67ab6d252125c129ed7ef84df833cb8f24082/dash/development/build_process.py#L153) relied on the `dash-renderer` package being installed and present in the environment. Since that package will no longer exist, I've moved the `dash-renderer` directory into the `dash` package so it's added to the source dist and the existing scripts work in line with it. 

>My rationale here is that as you have it here we're duplicating files: the ones actually getting used by dash are going to be harder to keep updated during renderer development because they need to go through an extra step, and the ones being created inside the renderer directory never get directly used at all. Also dash_renderer/__init__.py (based on templating from init.template) is a little cleaner than the new _dash_renderer.py, as it has all the versions inlined.

Updated in 9f8e0aa, with slight modifications to the existing template to have the correct relative url. 

>Finally (and again this is thinking about development in the renderer itself) since the entire dash/deps directory is build artifacts based on source files elsewhere in the repo, it should be gitignored except in the released (master) branch.

I'm a little confused on this point - presently, we can `gitignore` the dependencies in the existing `dash_renderer` directory because they'll be sourced from the `dash-renderer` package which is installed alongside `dash`. However, when that it will no longer exist, I think we might need to leave these in the `deps` directory in order to directly source them from within `dash`, or the `dev` branch version will fail to import it.

alexcjohnson commented:
> I think we might need to leave these in the deps directory in order to directly source them from within dash, or the dev branch version will fail to import it.

That's right, the committed code in dev branches will not run on its own. In the dev workflow, it's normal that after checking out the branch you want, you need to run the build command to create the relevant bundles & artifacts before you can actually run dash.

HammadTheOne commented:
@alexcjohnson I've made some changes to the CI config so that the artifacts are built within the `dash` package before installing and running tests and those are passing now. Unfortunately I don't have permissions to approve the Percy snapshots but it does look fine to me as well.

Other than that, I think this PR might be ready for a final pass, if the dependency structure makes sense at this point.

jonmmease commented:
@alexcjohnson @HammadTheOne  Is there anything else we need to pull the trigger on this one?

alexcjohnson commented on MANIFEST.in:
```diff
@@ -3,3 +3,6 @@ include LICENSE
 include dash/favicon.ico
 include dash/extract-meta.js
+include dash/deps/*.js
+recursive-include dash/dash-renderer *
```
This will include all the source files, won't it? I don't think we want to do that. The old dash-renderer `MANIFEST.in` is just:
```
include package.json
include digest.json
include dash_renderer/*.js
include dash_renderer/*.map
```
In the new structure, `dash_renderer/*.js` is replaced by `dash/deps/*.js` which you have above. I'm not aware of any value that `package.json` or `digest.json` have to end users so we should be able to ignore those...

Then there's the question of the `*.map` files: you're not currently including them, and in fact in the past in response to a user request to reduce the package size https://github.com/plotly/dash/issues/810 we at one point removed them https://github.com/plotly/dash/pull/910 but then for some reason that I can't find any record of we immediately reverted that https://github.com/plotly/dash/commit/369ace90b4af72d8ace1daaae9479b1129a462ee before making a release. @Marc-Andre-Rivet do you have any memory of that? I don't see a PR or any discussion, just a revert commit

Anyway, is there anything else we really do want from the dash-renderer directory, or can we drop this line (as well as the line below that only exists in reference to this one)?

HammadTheOne replied:
Updated in afc363cc. We can safely drop the other lines

I'm noticing that `.map` files aren't being generated by the webpack for souce-mapping. Is this because we have `inline-source-map`?

alexcjohnson replied:
Ah hmm, actually this changed in #1576 in a way that I hadn't recognized - @Marc-Andre-Rivet please confirm I'm understanding this correctly: the ONLY case we're generating a sourcemap now is when you run `webpack-serve` (which is what you get with `npm start`). The other build commands now seem to be all the same (as far as the bundle itself is concerned) and create no sourcemap. `npm run build:js` and `build:dev` are now literally the same command.

Unfortunately I can't actually run `npm start` ATM, and I don't think it does what we want anyway - really I'd like a command that outputs the same files we currently create, but stays open watching the source files for changes. @Marc-Andre-Rivet didn't we have that in the renderer at some point? Am I missing something about the updated tooling?

alexcjohnson commented on dash/development/build_process.py:
```diff
@@ -33,7 +33,7 @@ def _parse_package(self, path):
             self.version = package["version"]
             self.name = package["name"]
-            self.build_folder = self._concat(self.main, self.name.replace("-", "_"))
+            self.build_folder = self._concat(os.getcwd(), os.pardir, "deps")
```
Can we avoid involving the CWD here, and do something relative to `self.main`? Then we should be able to avoid `cd`ing into a certain directory before calling `renderer build`, and have it work from any directory.

HammadTheOne replied:
One problem I'm having with using `self.main` to set a relative path is that `build_process` is called from and points to the `site-packages/dash/dash-renderer` directory in the Python environment, rather than the development directory. So as a workaround I've used `os.getcwd()` so the build artifacts are properly updated into the working directory that the package is being developed in. I'm not entirely sure if that was different prior to `dash-renderer` being merged partially with `dash`. 

Could we maybe change that by calling `build_process` in the pwd rather than with a `console_scripts` entrypoint?

alexcjohnson replied:
Do you have dash installed with `-e`? ie `pip install -e .`? That use the dev directory as the installed version.

HammadTheOne replied:
:man_facepalming: Ah, that's what it was, I think somewhere between commits I reinstalled it as a tarfile. I'll have that updated shortly.

alexcjohnson commented on dash/__init__.py:
```diff
@@ -9,3 +9,4 @@
 from . import resources  # noqa: F401,E402
 from .version import __version__  # noqa: F401,E402
 from ._callback_context import callback_context  # noqa: F401,E402
+from ._dash_renderer import _js_dist_dependencies, _js_dist  # noqa: F401,E402
```
Do we need this import? This file is generally for items we want to expose to end users, these are imported directly by `dash.py` where they're used.

HammadTheOne replied:
Removed in 2aefb7fc.

alexcjohnson commented on .pylintrc:
```diff
@@ -62,7 +62,8 @@ disable=fixme,
         line-too-long,
-        bad-option-value
+        bad-option-value,
+        assigning-non-slot
```
why do we need to add `assigning-non-slot` to the disabled list?

alexcjohnson replied:
OIC this is due to `flask.g.*** =` and we're getting this error in all new CI runs now, eg [dependabot PRs](https://app.circleci.com/pipelines/github/plotly/dash/2245/workflows/412fba8d-f052-4de3-9fdc-23c14680608d/jobs/34927) and this is a result of changes in Flask v2.0 (👋 @anders-kiaer https://github.com/pallets/flask/issues/4020)

The good news is Dash works fine with Flask 2.0 🎉 

Rather than a global override of that rule, can we just skip it on the relevant lines with `# pylint: disable=assigning-non-slot` as in https://github.com/equinor/webviz-config/pull/451 ?

anders-kiaer replied:
> The good news is Dash works fine with Flask 2.0 🎉

Indeed 👍 - we were also glad to see Dash out of the box worked with Flask 2.0 🚀. 

Off topic: Slightly related to the `pylint` issue here - I _think_ Flask started using `__slots__` without `__dict__` on the `g` object for further performance improvements. I started wondering if using `__slots__` could be a useful and simple improvement idea for Dash as well (#1657) - not sure (I have no experience using `__slots__` directly myself as of today).

alexcjohnson commented on .circleci/config.yml:
```diff
@@ -256,12 +257,7 @@ jobs:
-      - run:
-          name: 🔧 fix up dash metadata
-          command: |
-            sudo Rscript dashR/tests/circleci/fixup_metadata.R
```
This step is no longer necessary?

HammadTheOne replied:
That's correct. The purpose of [this script](https://github.com/plotly/dashR/blob/master/tests/circleci/fixup_metadata.R) was to clean up versioning errors and update the latest commit hashes for `dcc`, `html`, and `table`. But now that those are included within `dashR` this is no longer necessary. There's some other cleanup we can do for other scripts involving `dashR` as well, but I think that's worth it's own PR. This change however, is necessary to get the tests to run.

alexcjohnson reviewed (approved):
Excellent! 💃  once there's an issue open to follow up on sourcmaps https://github.com/plotly/dash/pull/1611#discussion_r641143808
